rebasehelper/plugins/spec_hooks/replace_old_version.py

```diff
@@ -58,22 +61,54 @@ def _is_local_source(cls, line: str) -> bool:
         return not urllib.parse.urlparse(source).scheme
 
     @classmethod
-    def run(cls, spec_file, rebase_spec_file, **kwargs):
+    def _create_possible_replacements(cls, old: str, new: str, replacement: str) -> Dict[Pattern[str], str]:
+        """Creates possible subversion replacements.
+
+        Args:
+            old: Old version.
+            new: New Version.
+            replacement: Regex replacement pattern.
+
+        Returns:
+            Mapping of old version substring regex Pattern to replacement
+            for re.sub.
+
+        Example:
+            Subversions 1.2, 1.2.3 would be created from version 1.2.3.4.
+
+        """
+        split_version = old.split('.')
+        res = {}
+        # iterate backwards to go from longer to shorter subversions
+        for i in reversed(range(2, len(split_version))):
+            pattern = re.compile(r'([/\-\s]){}([/.\-\s])'.format(re.escape('.'.join(split_version[:i]))))
+            res[pattern] = replacement.format('.'.join(new.split('.')[:i]))
+        return res
+
+    @classmethod
+    def run(cls, spec_file: SpecFile, rebase_spec_file: SpecFile, **kwargs: Any):
         old_version = spec_file.get_version()
         new_version = rebase_spec_file.get_version()
         replace_with_macro = kwargs.get('replace_old_version_with_macro')
-        pattern = re.compile(r'([/\-\s]){}([/.\-\s])'.format(re.escape(old_version)))
+
+        replacement_pattern = r'\g<1>{}\g<2>'
+        subversion_patterns = cls._create_possible_replacements(old_version, new_version, replacement_pattern)
+
+        main_pattern = re.compile(r'([/\-\s]){}([/.\-\s])'.format(re.escape(old_version)))
```

Comment:
This shouldn't be special-cased, it should be part of the dictionary of patterns.

Reply:
I've considered this, the question is - do we want to try to replace subversions everywhere, not just in remote Sources/Patches? Couldn't there be any drawbacks? With this implementation, the full version is replaced everywhere (except for `%changelog`) and the version substrings only in Sources and Patches.

Regarding the next comment - if we opted for having the main replacement pattern in the dictionary, then it would be possible to define the replacement in `_create_pssible_replacements()`.

Reply:
That's a good point, but for example you can return a list of (pattern, replacement) tuples, with the first one always being the entire version.

Reply:
Good idea, that should be more understandable, because all replacements will be created in one place.